sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,315 @@
+# -*- coding: utf-8 -*-
+# Author: Arturo Amor
+# License: 3-clause BSD
+"""
+Recommendation system generator
+===============================
+
+Generate recommendations based on TF-IDF representation and a KNN model.
+"""
+# %%
+import numbers
+# import pickle
+import re
+from collections import defaultdict
+from pathlib import Path
+
+import numpy as np
+from scipy import sparse
+from scipy.sparse.linalg import norm
+
+from .backreferences import (
+    _thumbnail_div,
+    THUMBNAIL_PARENT_DIV,
+    THUMBNAIL_PARENT_DIV_CLOSE,
+)
+from .py_source_parser import split_code_and_text_blocks
+from .gen_rst import extract_intro_and_title
+
+
+class ExampleRecommender:
+    """Compute content-based KNN-TF-IFD recommendation system.
+
+    Parameters
+    ----------
+    n_examples : int, default=5
+        Number of most relevant examples to display.
+
+    tokenizer : {"raw", "backrefs"}, default="raw"
+        The type of tokenizer to use. If "raw", the raw text will be used as
+        tokens. If "backrefs", the list of sphinx-gallery backreferences will
+        be used as tokens.
+
+    Attributes
+    ----------
+    file_names_ : list of str
+        The list of file names used for computing the similarity matrix.
+        The recommended examples will be chosen among this list.
+
+    similarity_matrix_ : sparse matrix
+        Fitted matrix of pairwise cosine similarities.
+    """
+
+    def __init__(self, *, n_examples=5, tokenizer="raw"):
+        self.n_examples = n_examples
+        self.tokenizer = tokenizer
+
+    @staticmethod
+    def token_freqs(doc):
+        """Extract a dict mapping raw tokens from doc to their occurrences."""
+        token_generator = (tok.lower() for tok in re.findall(r"\w+", doc))
+        freq = defaultdict(int)
+        for tok in token_generator:
+            freq[tok] += 1
+        return freq
+
+    @staticmethod
+    def dict_freqs(doc):
+        """Extract a dict mapping list of tokens to their occurrences."""
+        freq = defaultdict(int)
+        for tok in doc:
+            freq[tok] += 1
+        return freq
+
+    @staticmethod
+    def dict_vectorizer(data_dict):
+        """Convert a dictionary of feature arrays into a sparse matrix.
+
+        Parameters
+        ----------
+        data_dict : dict
```

Comment:
This does not look like a `dict` but a `list of dict` from what I see in the code, isn't it?